Allow overwriting the path to a safe file

leonklingele opened the pull request.

leonklingele commented:
Requires #172

leonklingele commented:
@bndw that sums it up, updated commit message to be a bit more descriptive. Rebased to `develop`.

I was unsure whether `--file` really is the best name for the flag. What do you think? Some other options:

- `--safe`
- `--safefile`
- `--pickfile`

bndw commented:
@leonklingele I think `--safe` is a bit more intuitive. Since pick uses `~/.pick` as its config directory, would it make sense to support safe files relative to that directory? Perhaps we could support both?

e.g.
```
pick --safe work.safe ls
```

instead of 
```
pick --safe ~/.pick/work.safe ls
```

leonklingele commented:
Renamed `--file` -> `--safe` and rebased to `develop`.

The reason I did not choose to add support for pick-folder relative safe files is that I would expect
`pick --safe work.safe` to look for a safe `work.safe` in the __current working directory__, not in __`~/.pick/`__.

To support both options, we would need to distinguish between `--safe work.safe` and `--safe ./work.safe` (notice the leading `./`):

- `pick --safe work.safe` will look for `work.safe` in `~/.pick/`
- `pick --safe ./work.safe` will look for `work.safe` in the CWD

What do you think?

bndw commented:
I think that sounds good. I'll expand on it a bit for all backends:

### File
```
pick --safe work.safe will look for work.safe in ~/.pick/
```
```
pick --safe ./work.safe will look for work.safe in the CWD
```

### S3
```
pick --safe work.safe will look for work.safe in storage.settings.bucket
```
Relative safes in the remote S3 backend are handled the same (there is no relative or CWD for a remote backend)
```
pick --safe ./work.safe will look for work.safe in storage.settings.bucket
```

leonklingele commented:
@bndw implemented, ptal

leonklingele commented:
I haven't tested the changes to the s3 backend.

leonklingele commented on backends/s3/new.go:
```diff
@@ -25,6 +26,7 @@ func _new(config *backends.Config) (backends.Client, error) {
 	if !ok {
 		bucket = defaultS3Bucket
 	}
+	bucket = path.TrimModPrefix(bucket)
```
Oops, then I misunderstood your explanation in https://github.com/bndw/pick/pull/173#issuecomment-379340110. Does the following patch represent what you want me to change?

```patch
diff --git a/backends/s3/new.go b/backends/s3/new.go
index ba3c582..8306cea 100644
--- a/backends/s3/new.go
+++ b/backends/s3/new.go
@@ -26,12 +26,12 @@ func _new(config *backends.Config) (backends.Client, error) {
 	if !ok {
 		bucket = defaultS3Bucket
 	}
-	bucket = path.TrimModPrefix(bucket)
 
 	key, ok := config.Settings["key"].(string)
 	if !ok {
 		key = defaultS3Key
 	}
+	key = path.TrimModPrefix(key)
 
 	// TODO(bndw): Consider creating the bucket if it does not exist
 
```

I'm a complete S3 beginner, so please forgive my newbness. Should the `key` be prefixed with `safes/` as done for the default safe key?

bndw replied:
My mistake, I definitely said bucket in that comment.

An S3 bucket can be _loosely_ thought of as a directory that can hold any number of files/folders. I think it's most intuitive for the `--safe` flag to specify _which safe_ to use in the bucket (directory).

I believe the above patch would achieve this.

leonklingele replied:
What about the `safes/` prefix? (see above)

bndw replied:
Not adding the `safes/` prefix makes sense to me, though this may be inconsistent. 

Here's an example of how I configure this today and how I would use the `--safe` flag:

```toml
# ~/.pick/config.toml

[storage]
  type = "s3"

  [storage.settings]
    bucket  = "my-s3-bucket"
    key     = "safes/pick.safe"
    region  = "us-west-2"
    profile = "personal"
```

```sh
# List passwords in my `pick.safe`
pick ls

# List passwords in a safe in a different directory structure
pick --safe work/safes/pick.safe ls
```

leonklingele replied:
Updated with the patch posted before. This will however not allow specifying the bucket name, is that something we really want to support?

bndw replied:
In my usage I wouldn't ever change the bucket, so I think it's fine to only allow that in the config. If someone requests it as a flag we can go from there :)
